### PR TITLE
fix: move array_except in SetOp and support Null columnar in `array_except`, `array_union` and `array_intersect`

### LINKED ISSUE #9706: `array_union` and `array_intersect` cannot handle NULL columnar data
### Describe the bug

It seems that the functions `array_union` and `array_intersect` are unable to process columnar data that contains NULL values. These NULL values are being overlooked in the process.

### To Reproduce

``` bash
❯ 
CREATE TABLE array_intersect_table
AS VALUES
  ([1, 2, 2, 3], [2, 3, 4]),
  ([2, 3, 3], [3]),
  ([3], [3, 3, 4]),
  (null, [3, 4]),
  ([1, 2], null),
  (null, null)
;
0 rows in set. Query took 0.013 seconds.

❯ select array_intersect(column1, column2) from array_intersect_table;
+------------------------------------------------------------------------+
| array_intersect(array_intersect_table.column1,array_except_table.column2) |
+------------------------------------------------------------------------+
| [2, 3]                                                                 |
| [3]                                                                    |
| [3]                                                                    |
+------------------------------------------------------------------------+
3 rows in set. Query took 0.007 seconds.
```

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Weijun-H**
## Which issue does this PR close?

Closes #9706

## Rationale for this change
- move `array_except` to  SetOp 
- support Null columnar for array set opertaion

### CODE REVIEW

File: datafusion/functions-array/src/set_ops.rs [hunk 1]
```diff
@@ -273,55 +284,97 @@ fn generic_set_lists<OffsetSize: OffsetSizeTrait>(
...
+            Some(offset) => offset,
+            None => return internal_err!("offsets should not be empty"),
+        };
+        match (first_arr, second_arr) {
+            (Some(first_arr), Some(second_arr)) => {
+                let mut l_values = converter.convert_columns(&[first_arr])?;
+                let mut r_values = converter.convert_columns(&[second_arr])?;
+
+                if set_op == SetOp::Except {
+                    (l_values, r_values) = (r_values, l_values);
```
Comment (jayzhan211) [inline datafusion/functions-array/src/set_ops.rs:301]: I think mixing the logic of except here make it less readable 😕

File: datafusion/sqllogictest/test_files/array.slt [hunk 2]
```diff
@@ -405,6 +405,18 @@ AS VALUES
   (arrow_cast(make_array([5,6], [5,6], NULL, NULL, NULL), 'FixedSizeList(5, List(Int64))'))
 ;
 
+statement ok
+CREATE TABLE array_setop_with_NULL_table
```
Comment (jayzhan211) [inline datafusion/sqllogictest/test_files/array.slt:409]: you can name it `array_setop`. for all the setop functions

**Comment by jayzhan211** [comment]
@Weijun-H I think the common logic for them is not large enough to worth mixing into one function. Do you think separate the function for them is easier for understanding the code?  I'm even thinking of `union` and `intersect` too.

File: datafusion/functions-array/src/set_ops.rs [hunk 3]
```diff
@@ -273,55 +284,88 @@ fn generic_set_lists<OffsetSize: OffsetSizeTrait>(
...
+                    if values_set.contains(&r_val) {
+                        rows.push(r_val);
+                    }
+                }
+            }
         }
+
+        offsets.push(last_offset + OffsetSize::usize_as(rows.len()));
+        let arrays = converter.convert_rows(rows)?;
+        let array = match arrays.first() {
```
Comment (jayzhan211) [inline datafusion/functions-array/src/set_ops.rs:350]: I recently learn `swap_remove`, I think we can apply here without clone. `arrays.swap_remove(0)`

**Comment by jayzhan211** [comment]
How about extract the common to function. And, try to let the logic for one
stay close together.

One of issue I think that make it less readable is that for example array
except, we need to go through several if else statements that is unrelated
(union or intersect), and even swapping for the common function.

I think I need suggestions from others too, maybe it is easy to read for
others.

**Comment by alamb** [comment]
Marking as draft as I think this PR is no longer waiting on feedback. Please mark it as ready for review when it is ready for another look

**Pull request closed without merging**